nixos/modules/system/boot/luksroot.nix

```diff
@@ -1096,42 +1096,33 @@ in
     boot.initrd.postDeviceCommands = mkIf (!config.boot.initrd.systemd.enable) (commonFunctions + preCommands + concatStrings (mapAttrsToList openCommand postLVM) + postCommands);
 
     boot.initrd.systemd.services = let devicesWithClevis = filterAttrs (device: _: (hasAttr device clevis.devices)) luks.devices; in
-      mkIf (clevis.enable && systemd.enable) ((mapAttrs'
-        (name: _: nameValuePair "cryptsetup-pre-clevis-${name}" {
-          requiredBy = [ "systemd-cryptsetup@${name}.service" "cryptsetup.target" "cryptsetup-pre.target" ];
-          before = [ "cryptsetup.target" "cryptsetup-pre.target" ];
-          wants = [ "systemd-udev-settle.service" "network-online.target" ];
-          after = [
-            "systemd-udev-settle.service"
-            "systemd-modules-load.service"
-          ];
-          unitConfig.DefaultDependencies = "no";
-          serviceConfig = {
-            Type = "oneshot";
-            RemainAfterExit = true;
-          };
-          script = ''
-            mkdir -p /clevis-${name}
-            mount -t ramfs none /clevis-${name}
-            clevis decrypt < /etc/clevis/${name}.jwe > /clevis-${name}/decrypted
-          '';
-        })
-        devicesWithClevis) // (mapAttrs'
-        (name: _: nameValuePair "cryptsetup-post-clevis-${name}" {
-          requires = [ "cryptsetup.target" ];
-          after = [
-            "cryptsetup.target"
-            "systemd-udev-settle.service"
-            "systemd-modules-load.service"
-            "systemd-ask-password-console.service"
-          ];
-          unitConfig.DefaultDependencies = "no";
-          serviceConfig.Type = "oneshot";
-          script = ''
-            umount /clevis-${name}
-          '';
-        })
-        devicesWithClevis));
+      mkIf (clevis.enable && systemd.enable) (
+        (mapAttrs'
+          (name: _: nameValuePair "cryptsetup-clevis-${name}" {
+            requiredBy = [ "systemd-cryptsetup@${name}.service" ];
+            before = [ "systemd-cryptsetup@${name}.service" ];
+            wants = [ "systemd-udev-settle.service" "network-online.target" ];
+            after = [
+              "systemd-udev-settle.service"
+              "systemd-modules-load.service"
+              "network-online.target"
+            ];
+            script = ''
+              mkdir -p /clevis-${name}
+              mount -t ramfs none /clevis-${name}
+              clevis decrypt < /etc/clevis/${name}.jwe > /clevis-${name}/decrypted
+              chmod 400 /clevis-${name}/decrypted
+            '';
+            conflicts = [ "initrd-switch-root.target" "shutdown.target" ];
```

Comment:
There also needs to be an ordering between this unit and these conflicting ones. It doesn't matter if the ordering is before or after (though before is typical); when there is *any* ordering between two units, and one is being stopped at the same time that the other is being started, the stop happens before the start. Without the ordering, however, the stop can happen simultaneously, which is especially problematic when isolating something like `initrd-switch-root.target`.

Reply:
Thank you, that makes sense :)